presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PhysicalResourceOptimizer.java

```diff
@@ -0,0 +1,308 @@
+                    calculateLatencyOptimizedExecutorCountForShorterQueries((int) numberOfExecutors);
+        }
+    }
+
+    private int calculateLatencyOptimizedExecutorCountForShorterQueries(int numberOfExecutors)
+    {
+        int updatedExecutorCount = Math.max(numberOfExecutors, MIN_EXECUTOR_FOR_SMALL_QUERIES_LATENCY_OPTIMIZED);
+        int finalExecutorCount = Math.max(MIN_EXECUTOR_FOR_SMALL_QUERIES_LATENCY_OPTIMIZED, Math.min(MAX_EXECUTOR_FOR_SMALL_QUERIES_LATENCY_OPTIMIZED, updatedExecutorCount));
+        return finalExecutorCount;
+    }
+
+    private int calculateDircuOptimizedExecutorCountForShorterQueries(int numberOfExecutors)
+    {
+        int updatedExecutorCount = Math.max(numberOfExecutors, MIN_EXECUTOR_FOR_SMALL_QUERIES_DIRCU_OPTIMIZED);
+        int finalExecutorCount = Math.max(MIN_EXECUTOR_FOR_SMALL_QUERIES_DIRCU_OPTIMIZED, Math.min(MAX_EXECUTOR_FOR_SMALL_QUERIES_DIRCU_OPTIMIZED, updatedExecutorCount));
+        return finalExecutorCount;
+    }
+
+    private int calculateDircuOptimizedExecutorCountForLongerQueries(int numberOfExecutors)
+    {
+        int updatedExecutorCount = Math.max(numberOfExecutors, MIN_EXECUTOR_FOR_LARGE_QUERIES_DIRCU_OPTIMIZED);
+        int finalExecutorCount = Math.max(MIN_EXECUTOR_FOR_LARGE_QUERIES_DIRCU_OPTIMIZED, Math.min(MAX_EXECUTOR_FOR_LARGE_QUERIES_DIRCU_OPTIMIZED, updatedExecutorCount));
+        return finalExecutorCount;
+    }
+
+    private int calculateLatencyOptimizedExecutorCountForLongerQueries(int numberOfExecutors)
+    {
+        int updatedExecutorCount = Math.max(numberOfExecutors, MIN_EXECUTOR_FOR_LARGE_QUERIES_LATENCY_OPTIMIZED);
+        int finalExecutorCount = Math.max(MIN_EXECUTOR_FOR_LARGE_QUERIES_LATENCY_OPTIMIZED, Math.min(MAX_EXECUTOR_FOR_LARGE_QUERIES_LATENCY_OPTIMIZED, updatedExecutorCount));
+        return finalExecutorCount;
+    }
+    public class CollectSourceStats
+    {
+        private final Metadata metadata;
+        private final Session session;
+
+        public CollectSourceStats(Metadata metadata, Session session)
+        {
+            this.metadata = metadata;
+            this.session = session;
+        }
+
+        public AbstractMap.SimpleImmutableEntry<Double, Boolean> collectSourceStats(PlanNode root)
+        {
+            Visitor visitor = new Visitor();
+            root.accept(visitor, null);
+
+            AbstractMap.SimpleImmutableEntry<List<TableStatistics>, Boolean> sourceStatistics = visitor.getStatsForSources();
+
+            Iterator<TableStatistics> tableStatisticsIterator = sourceStatistics.getKey().iterator();
+            Double totalSourceSize = 0.0;
+            while (tableStatisticsIterator.hasNext()) {
+                TableStatistics ts = tableStatisticsIterator.next();
+                totalSourceSize = totalSourceSize + ts.getTotalSize().getValue();
+            }
+
+            return new AbstractMap.SimpleImmutableEntry(totalSourceSize, sourceStatistics.getValue());
```

Comment:
It seems that if `isSourceMissingData` is true, we will skip the tuning. But here, we are computing the total sizes and sending it back to caller where this flag is being checked. We can probably check this flag first and return -1 as the datasize immediately which indicates that the source data is missing.  Then in caller code, you can just check that if dataSize >= 0, then only perform the tuning.

Reply:
Simplified the logic and moved the dataSize check to the caller side as per feedback.